drivers/flash/flash_stm32h7x.c

```diff
@@ -469,21 +537,52 @@ flash_stm32h7_get_parameters(const struct device *dev)
 	return &flash_stm32h7_parameters;
 }
 
-static const struct flash_pages_layout stm32h7_flash_layout[] = {
-	{ .pages_count = FLASH_SECTOR_TOTAL, .pages_size = FLASH_SECTOR_SIZE },
-#ifdef DUAL_BANK
-	{ .pages_count = FLASH_SECTOR_TOTAL, .pages_size = FLASH_SECTOR_SIZE },
-#endif
-};
 
 void flash_stm32_page_layout(const struct device *dev,
 			     const struct flash_pages_layout **layout,
 			     size_t *layout_size)
 {
 	ARG_UNUSED(dev);
 
-	*layout = stm32h7_flash_layout;
+#if defined(DUAL_BANK)
+	static struct flash_pages_layout stm32h7_flash_layout[3];
+
+	if (DISCONTINUOUS_BANKS) {
+		if (stm32h7_flash_layout[0].pages_count == 0) {
+			/* Bank1 */
+			stm32h7_flash_layout[0].pages_count = SECTOR_PER_BANK;
+			stm32h7_flash_layout[0].pages_size = FLASH_SECTOR_SIZE;
+			/*
+			 * Dummy page corresponding to discontinuity
+			 * between bank1/2
+			 */
+			stm32h7_flash_layout[1].pages_count = 1;
+			stm32h7_flash_layout[1].pages_size = BANK2_OFFSET
+					- (SECTOR_PER_BANK * FLASH_SECTOR_SIZE);
+			/* Bank2 */
+			stm32h7_flash_layout[2].pages_count = SECTOR_PER_BANK;
+			stm32h7_flash_layout[2].pages_size = FLASH_SECTOR_SIZE;
+		}
+		*layout_size = ARRAY_SIZE(stm32h7_flash_layout);
+	} else {
+		if (stm32h7_flash_layout[0].pages_count == 0) {
+			stm32h7_flash_layout[0].pages_count =
+				REAL_FLASH_SIZE_KB / FLASH_SECTOR_SIZE;
+			stm32h7_flash_layout[0].pages_size = FLASH_SECTOR_SIZE;
+		}
+		*layout_size = 1;
+	}
+#else
+	static struct flash_pages_layout stm32h7_flash_layout[1];
+
+	if (stm32h7_flash_layout[0].pages_count == 0) {
+		stm32h7_flash_layout[0].pages_count =
+				REAL_FLASH_SIZE / FLASH_SECTOR_SIZE;
+		stm32h7_flash_layout[0].pages_size = FLASH_SECTOR_SIZE;
+	}
 	*layout_size = ARRAY_SIZE(stm32h7_flash_layout);
+#endif
```

Comment:
```suggestion
#if defined(DUAL_BANK)
#define FLASH_BANKS_NUM		3  // including one dummy bank
#else  // DUAL_BANK
#define FLASH_BANKS_NUM		1
#endif  // DUAL_BANK

	static struct flash_pages_layout stm32h7_flash_layout[FLASH_BANKS_NUM];

	if (DISCONTINUOUS_BANKS) {
		if (stm32h7_flash_layout[0].pages_count == 0) {
			/* Bank1 */
			#if defined(DUAL_BANK)
			stm32h7_flash_layout[0].pages_count = SECTOR_PER_BANK;
			#else  // DUAL_BANK
			stm32h7_flash_layout[0].pages_count =
				REAL_FLASH_SIZE_KB / FLASH_SECTOR_SIZE;
			#endif  // DUAL_BANK

			stm32h7_flash_layout[0].pages_size = FLASH_SECTOR_SIZE;
			/*
			 * Dummy page corresponding to discontinuity
			 * between bank1/2
			 */
			stm32h7_flash_layout[1].pages_count = 1;
			stm32h7_flash_layout[1].pages_size = BANK2_OFFSET
					- (SECTOR_PER_BANK * FLASH_SECTOR_SIZE);
			/* Bank2 */
			stm32h7_flash_layout[2].pages_count = SECTOR_PER_BANK;
			stm32h7_flash_layout[2].pages_size = FLASH_SECTOR_SIZE;
		}
		*layout_size = ARRAY_SIZE(stm32h7_flash_layout);
	} else {
		if (stm32h7_flash_layout[0].pages_count == 0) {
			stm32h7_flash_layout[0].pages_count =
				REAL_FLASH_SIZE_KB / FLASH_SECTOR_SIZE;
			stm32h7_flash_layout[0].pages_size = FLASH_SECTOR_SIZE;
		}
		#if defined(DUAL_BANK)
		*layout_size = 1;
		#else  // DUAL_BANK
		*layout_size = ARRAY_SIZE(stm32h7_flash_layout);
		#endif  // DUAL_BANK
	}
	*layout = stm32h7_flash_layout;
}

```

Above suggestion should produce less code.

Reply:
This may be true on some cases (few code line saved), but false on other:
In case of Single bank, the code is much larger due to `if (DISCONTINUOUS_BANKS)`  part which is  quite large and is a non sense for single bank (and may introduce confusion).
So I prefer to keep like this.